### DISCUSSION TITLE
Syncing Shared Business Files

### DISCUSSION THREAD

**Discussion by AnthonyAnkrah** [abraunegg/onedrive#2644] [Q&A] [answered]
Hi community, I have `v2.4.25-13` of the onedrive client running in a docker container. 
![version](https://github.com/abraunegg/onedrive/assets/59777079/6dc729da-a2ae-49de-9513-03d93173a403)

I have setup a `business_shared_folders` file and specified the root folders to sync (shown below).
![business_shared_folders](https://github.com/abraunegg/onedrive/assets/59777079/ee72ffc4-7728-4bcd-8f97-ddd86fb6a6ac)

I also have a `sync_list` file that specifies the subdirectories of these shared folders that need to be synced (also shown below)
![sync_list](https://github.com/abraunegg/onedrive/assets/59777079/be43845a-2b59-4c54-b3ee-d43ddd5fa1fd)

The directories specified on lines 3 and 4 of the `sync_list` file are syncing just fine and I have no issues there. However, I have a file (not a directory) that is also shared with that same account that I would like to sync as well. As shown above in both screenshots, I have included the filename in both the `business_shared_folders` and `sync_list` files but can't get the file to sync for the life of me. Any suggestions on how I could accomplish this?
![shared file](https://github.com/abraunegg/onedrive/assets/59777079/3e50e68a-3521-4b17-ab4f-72854cb2807c)

Also, I specified a custom location for logs and attached it to the container but that directory is consistently empty. Running a `onedrive --display-config` command while in the container shows a different set of config options that do not include the values I specified despite all other things running as intended. Am I missing something here?
![config](https://github.com/abraunegg/onedrive/assets/59777079/6165c590-da25-4471-85ad-277486372215)

![docker-compose](https://github.com/abraunegg/onedrive/assets/59777079/f4da145e-0209-4fd5-b333-082f1645d6c1)

**Comment by abraunegg** [comment, marked as answer]
@AnthonyAnkrah 
The syncing of business shared folders is just that - the syncing of folders, not files.

Additionally this is already a feature request - https://github.com/abraunegg/onedrive/issues/1300 .. however I have so far been unable to even share a single file to a user - only a folder .. so curious how this was even done as the OneDrive Web Ui provides zero capability to share an individual file.

**Reply by AnthonyAnkrah** [reply]
@abraunegg
Thanks for your timely response. Actually, there are 2 ways (that I know) to share a file with someone; via the web UI and from file explorer if you already have a local sync on your Windows PC. Below are 2 screenshots to help shed more light on the subject.
![share file-web](https://github.com/abraunegg/onedrive/assets/59777079/b2b7a277-72b5-4fba-a5c4-6e10643388f1)

![share file-file explorer](https://github.com/abraunegg/onedrive/assets/59777079/379028a8-0ee3-432e-adff-0cccaa32d8f0)

**Reply by abraunegg** [reply]
@AnthonyAnkrah 
Interesting .. so 'files' are not being shared by the normal mechanism ... let me investigate this further as part of #1300 so that this can be added into v2.5.x / 'alpha-5'

**Comment by abraunegg** [comment]
@AnthonyAnkrah 
Re the second question regarding logging - the container is being started with passing in a parameter to specify the config directory. Unless you pass this is when using `--display-config` you will not get the actual running config.

You can also pass the docker variable to display the runtime config to the docker console output to double check what is being passed in.

Setting a log dir though - does not enable logging to a file - you must also enable logging.

**Comment by abraunegg** [comment]
@AnthonyAnkrah 
One final note - the feature you are currently using will 100% change with v2.5.x ... please read all the 'alpha-5' documentation

**Reply by AnthonyAnkrah** [reply]
Talk of piqued interest😀, the flash has nothing on me. LoL

**Comment by abraunegg** [comment]
@AnthonyAnkrah

As per #1300, his Feature Request has now been completed.

**Important:** Before you do *anything* else ... please first *read* the updated documentation for OneDrive Business Shared Items: https://github.com/abraunegg/onedrive/blob/Implement-Feature-Request-1300/docs/business-shared-items.md

The entire feature has been re-written, how it is implemented has been changed - hopefully the documentation is clear enough on how to configure this feature going forward.

If you want to *test* the PR, please use the instructions below:

```
git clone https://github.com/abraunegg/onedrive.git
cd onedrive
git fetch origin pull/2646/head:pr2646
git checkout pr2646
```

All feedback - positive or negative (or something else) is highly appreciated.

To run the PR, you need to run the client from the PR build directory:
```
./onedrive <any options needed>
```

To install the PR, you will need to perform `sudo make install` to install the PR version. 

When running the PR, your version should be: `v2.5.0-alpha-5 GitHub version: v2.4.25-93-gfde6f86` or greater

**Reply by AnthonyAnkrah** [reply]
Will be trying it out within the hour. Will keep you posted @abraunegg

**Reply by abraunegg** [reply]
@AnthonyAnkrah 
Any feedback ?

**Reply by AnthonyAnkrah** [reply]
Sorry for the delayed response @abraunegg.
Got a minute to work on this yesterday and I was faced with a peculiar issue.

So whenever I try to get into bash with `docker exec -it onedrive sh` I get a successful connection but get almost immediately kicked out with an `exit code 137`. My search revealed that was associated with a memory limitation and I'm still trying to figure that out.

It is also worth noting that this is only occurring on my `"playground"` setup and not on my `"main"` instance as that (main) is able to sustain a bash connection for way longer.

For anyone who might be interested:
`"main"`
> docker version 24.0.2 build cb74dfc
> docker compose version v2.18.1
> onedrive v2.4.25-13-g1a88d33

`"playground"`
> docker version 25.0.4 build 1a576c5
> docker compose version v2.24.7
> onedrive v2.4.25-13-g1a88d33

While writing this response, I have realized most synced files are being duplicated several times with a recurring string appended to the filenames. See below.
![duplicate sync](https://github.com/abraunegg/onedrive/assets/59777079/8031a52d-095f-4328-8450-518ab30769e1)

**Reply by AnthonyAnkrah** [reply]
@abraunegg 
Also found this in the logs. 
```
??:? @trusted void std.file.renameImpl(scope const(char)[], scope const(char)[], scope const(char)*, scope const(char)*) [0x7f81979c6362]
src/config.d:932 [0x55578f14f515]
src/util.d:44 [0x55578f14f2e6]
src/sync.d [0x55578f17b2a8]
src/sync.d:2622 [0x55578f171e95]
src/sync.d [0x55578f168e94]
src/sync.d:816 [0x55578f1640c2]
src/main.d:1896 [0x55578f1c7f08]
src/main.d:1663 [0x55578f1b9952]
??:? void rt.dmain2._d_run_main2(char[][], ulong, extern (C) int function(char[][])*).runAll() [0x7f8197662b7b]
??:? _d_run_main2 [0x7f8197662995]
??:? _d_run_main [0x7f81976627ed]
??:? [0x7f81972bf1c9]
??:? __libc_start_main [0x7f81972bf284]
??:? [0x55578f13aec0]
std.net.curl.CurlException@std/net/curl.d(4402): A libcurl function was given a bad argument on handle null
----------------
/usr/lib/ldc/x86_64-linux-gnu/include/d/std/exception.d:519 pure @safe noreturn std.exception.bailOut!(std.net.curl.CurlException).bailOut(immutable(char)[], ulong, scope const(char)[]) [0x55578f2a4cca]
??:? void std.net.curl.Curl.clear(etc.c.curl.CurlOption) [0x7f81979480ea]
src/onedrive.d:500 [0x55578f1a4e5f]
??:? void rt.dmain2._d_run_main2(char[][], ulong, extern (C) int function(char[][])*).runAll() [0x7f8197662b7b]
??:? _d_run_main2 [0x7f8197662995]
??:? _d_run_main [0x7f81976627ed]
??:? [0x7f81972bf1c9]
??:? __libc_start_main [0x7f81972bf284]
??:? [0x55578f13aec0]
usermod: no changes
Base Args: --monitor
# We are being verbose
# Adding --verbose
# We are performing a --resync
# Adding --resync --resync-auth
# Launching onedrive
Using 'user' Config Dir: /onedrive/conf
Error grabbing logs: invalid character '\x00' looking for beginning of value
```

**Reply by abraunegg** [reply]
@AnthonyAnkrah 
Your 'playground' is not even running the correct version of v2.5.x. Your screen shots also show clearly to me that this is also the case.

Given that RC-1 is now available, to build a docker container of v2.5.0 RC1:
```
git clone https://github.com/abraunegg/onedrive.git
cd onedrive
git fetch origin pull/2661/head:pr2661
git checkout pr2661
```
then, from inside the 'onedrive' clone that is pointing at RC1:
```
docker build . -t local-ondrive-debian -f contrib/docker/Dockerfile-debian
```

The next issue you will face is - you cannot run v2.4.25 Docker version and v2.5.0 Docker version at the same time - because of *how* the Shared Files Feature will be operating and what is needed to be done locally & online.

Please *do not* run the two Docker instances at the same time.

**Reply by AnthonyAnkrah** [reply]
@abraunegg 
I think you misunderstood. My `playground` and `main` setups are just that; 2 separate pieces of hardware. Thus, I am not running multiple versions of docker on the same system.

Also, I was reporting an issue being faced prior to rolling out the new fix. To update you on that issue, I pulled down both containers on the 2 systems to try identify which one was causing the file duplication problem. In the meantime, I manually removed all the duplicates from all the directories I was syncing. Sadly, after a good 5-day wait period, the duplicate files just sprung up the very moment I started the container on the main system.

Any pointers to help me diagnose this would be greatly appreciated.

**Reply by abraunegg** [reply]
You are running the wong client version & docker container on each system.

V2.4.x docker container and code is 100% incompatible with v2.5.x code

Do not run v2.4.x docker containers if you are using v2.5.x 

Next - enable logging to a file, enable config reporting for the docker containers, enable debug logging

**Reply by AnthonyAnkrah** [reply]
Oh okay. That's a revelation. The docker engine on the main system cannot be upgraded on account of other containers it is currently running. Which version of the client would you recommend I run on that system.

Also, I had already enabled logging to a file. Are there any specific sections you are interested in?

Also, I added the line below to my config file and spun the container back up (prior to seeing your last) but it didn't seem to take. 
```
skip_file = "~*|.~*|*.tmp|*/*ec7f1980c5fb*|*/*add7069fa800*"
```
Guess that might be tied to that incompatibility thingy?
Looking into the debug logging and config reporting next.
Cheers!

**Reply by abraunegg** [reply]
> Oh okay. That's a revelation. The docker engine on the main system cannot be upgraded on account of other containers it is currently running. Which version of the client would you recommend I run on that system.

If you are using v2.5.x anywhere .. ALL your onedrive containers MUST be v2.5.x ....

It is not backwards compatible

**Reply by AnthonyAnkrah** [reply]
Duly noted.
Now (more like for 5 days), I have the setup on the playground down. Any pointers on how I can get the setup on main working properly without re-generating those duplicate files?

**Reply by abraunegg** [reply]
Business Shared Files feature is ONLY a v2.5.0+ feature.

You cannot use this feature / configure aspects online and expect it to work with v2.4.x

**Reply by abraunegg** [reply]
@AnthonyAnkrah 
> Any pointers on how I can get the setup on main working properly without re-generating those duplicate files?

In the debug log - which is ONLY again using v2.4.25:

```
2024-Mar-21 00:46:00.8464711	[DEBUG] business_shared_folders: # Each line corresponds to a Business Shared Folder that should be synced
2024-Mar-21 00:46:00.8464964	[DEBUG] business_shared_folders: Approved Docs
2024-Mar-21 00:46:00.8465219	[DEBUG] business_shared_folders: GH-CNC.xlsx
```

You cannot specify files in `business_shared_folders` file .. it is FOLDERS only

I am also not seeing any 'duplicates' being created here ... do you have an actual example?

Again - please note - you asked about Business Shared Files. This is NOT a v2.4.25 feature. You MUST use v2.5.0 or above, and it is NOT backward compatible.

My suggestion here is that you clone the v2.5.0-rc1 codebase, manually build a Docker container from that codebase so you are running v2.5.0 as your Docker image.

So that this is clear on how you can achieve this:
```
git clone https://github.com/abraunegg/onedrive.git
cd onedrive
git fetch origin pull/2661/head:pr2661
git checkout pr2661
```
To then build the Docker container of v2.5.0-rc1, you need to follow the instructions listed here: https://github.com/abraunegg/onedrive/blob/onedrive-v2.5.0-release-candidate-1/docs/docker.md#building-and-running-a-custom-docker-image

You will have already done the 'clone' - so it is now the actual build:
* Build a Fedora based Docker image: `docker build . -t local-onedrive -f contrib/docker/Dockerfile`
* Build a Debian based Docker image: `docker build . -t local-ondrive-debian -f contrib/docker/Dockerfile-debian`
* Build a Alpine based Docker image: `docker build . -t local-ondrive-alpine -f contrib/docker/Dockerfile-alpine`

Once your Docker image is built, you can then 'run' the docker image as per the following:

* Fedora: `docker container run -v onedrive_conf:/onedrive/conf -v "${ONEDRIVE_DATA_DIR}:/onedrive/data" local-onedrive:latest`
* Debian: `docker container run -v onedrive_conf:/onedrive/conf -v "${ONEDRIVE_DATA_DIR}:/onedrive/data" local-ondrive-debian:latest`
* Alpine: `docker container run -v onedrive_conf:/onedrive/conf -v "${ONEDRIVE_DATA_DIR}:/onedrive/data" local-ondrive-alpine:latest`

**Reply by AnthonyAnkrah** [reply]
Hello again @abraunegg 
Thanks a ton for the above guide. It has been more useful that you could ever imagine.

Just wanted to draw your attention to some small thing. The environment variables listed in [this section](https://github.com/abraunegg/onedrive/blob/onedrive-v2.5.0-release-candidate-1/docs/docker.md#supported-docker-environment-variables) of the documentation does not include one for `--sync-shared-files` as captured in [this section](https://github.com/abraunegg/onedrive/blob/onedrive-v2.5.0-release-candidate-1/docs/business-shared-items.md#syncing-onedrive-business-shared-files) of the guide to syncing shared items. Would you recommend modifying the `entrypoint.sh` file?

**Reply by abraunegg** [reply]
@AnthonyAnkrah 
> Just wanted to draw your attention to some small thing. The environment variables listed in [this section](https://github.com/abraunegg/onedrive/blob/onedrive-v2.5.0-release-candidate-1/docs/docker.md#supported-docker-environment-variables) of the documentation does not include one for --sync-shared-files as captured in [this section](https://github.com/abraunegg/onedrive/blob/onedrive-v2.5.0-release-candidate-1/docs/business-shared-items.md#syncing-onedrive-business-shared-files) of the guide to syncing shared items. Would you recommend modifying the entrypoint.sh file?

Please raise a feature request for tracking: https://github.com/abraunegg/onedrive/issues/new?assignees=&labels=Feature+Request&projects=&template=feature_request.yml&title=Feature+Request%3A+

**Reply by AnthonyAnkrah** [reply]
This has been done in [#2680](https://github.com/abraunegg/onedrive/issues/2680)

**Reply by AnthonyAnkrah** [reply]
@abraunegg 
Been a while.

Pulled the update from commit #2686  and applied them on both the `main` and `playground` systems. Happy to report its working well.
👏🏽👏🏽
Thanks

**Reply by abraunegg** [reply]
@AnthonyAnkrah 
Please do not use #2686 .. that is currently a DEV branch where I am integrating #2608 int RC1 ... and I am finding lots of issues / problems with it. That it is *working* for you - great , my advice is do not use it at this point in time please.

You should *stick* with using #2661 - as this is the current RC1 build. 

Once all the issues are ironed out with #2686 - this will become RC2 in the next few weeks.

**Reply by AnthonyAnkrah** [reply]
Thanks for the heads-up @abraunegg 
Just reverted back to #2661
